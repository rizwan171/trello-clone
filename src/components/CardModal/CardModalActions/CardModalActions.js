import React from "react";
import { useDispatch } from "react-redux";
import { AiOutlineTag, AiOutlineDelete, AiOutlinePaperClip} from "react-icons/ai";
import { HiOutlineArrowRight } from "react-icons/hi";
import { MdOutlineContentCopy } from "react-icons/md";
import { useSelector } from "react-redux";
import { showCopyMenu, showDeleteMenu, showMoveMenu, showTagsMenu, showAttachmentMenu} from "../../../features/modalActionMenusVisibilitySlice";
import CreateTagMenu from "./TagsMenu/CreateTagMenu/CreateTagMenu";
import TagsMenu from "./TagsMenu/TagsMenu";
import CopyMenu from "./CopyMenu/CopyMenu";
import MoveMenu from "./MoveMenu/MoveMenu";
import DeleteMenu from "./DeleteMenu/DeleteMenu";
import AttachmentMenu from "./AttachmentMenu/AttachmentMenu";

const CardModalActions = () => {
  const dispatch = useDispatch();
  const tagsMenuOpen = useSelector((state) => state.modalActionMenusVisibility.value.tagsMenuOpen);
  const createTagsMenuOpen = useSelector((state) => state.modalActionMenusVisibility.value.createTagsMenuOpen);
  const copyMenuOpen = useSelector((state) => state.modalActionMenusVisibility.value.copyMenuOpen);
  const moveMenuOpen = useSelector((state) => state.modalActionMenusVisibility.value.moveMenuOpen);
  const deleteMenuOpen = useSelector((state) => state.modalActionMenusVisibility.value.deleteMenuOpen);
  const attachmentMenuOpen = useSelector((state) => state.modalActionMenusVisibility.value.attachmentMenuOpen);

  const handleShowTagsMenu = () => {
    dispatch(showTagsMenu());
  };

  const handleShowCopyMenu = () => {
    dispatch(showCopyMenu());
  };

  const handleShowMoveMenu = () => {
    dispatch(showMoveMenu());
  };

  const handleShowDeleteMenu = () => {
    dispatch(showDeleteMenu());
  };

  const handleShowAttachmentMenu = (id) => {
    dispatch(showAttachmentMenu(id));
  }

  return (
    <div className="flex flex-col w-44 py-4 ml-auto">
      <h4 className="text-gray-800 text-sm">Add to card</h4>
      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowTagsMenu}
        >
          <AiOutlineTag />
          <p>Tags</p>
        </button>
        {tagsMenuOpen && <TagsMenu card/>}
      </div>
      {createTagsMenuOpen && <CreateTagMenu/>}
      <h4 className="text-gray-800 text-sm">Actions</h4>
      
      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowCopyMenu}
        >
          <MdOutlineContentCopy />
          <p>Copy</p>
        </button>
        {copyMenuOpen && <CopyMenu />}
      </div>

      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowMoveMenu}
        >
          <HiOutlineArrowRight />
          <p>Move</p>
        </button>
        {moveMenuOpen && <MoveMenu />}
      </div>

      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowDeleteMenu}
        >
          <AiOutlineDelete />
          <p>Delete</p>
        </button>
        {deleteMenuOpen && <DeleteMenu />}
      </div>

      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={() => handleShowAttachmentMenu(1)}
        >
          <AiOutlinePaperClip />
          <p>Attachment</p>
        </button>
        {attachmentMenuOpen.status && attachmentMenuOpen.id===1 && <AttachmentMenu update={false}/>}
      </div>
    </div>
  );
};

export default CardModalActions;
