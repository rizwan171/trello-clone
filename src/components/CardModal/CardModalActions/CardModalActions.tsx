import { AiOutlineTag, AiOutlineDelete, AiOutlinePaperClip } from "react-icons/ai";
import { HiOutlineArrowRight } from "react-icons/hi";
import { MdOutlineContentCopy } from "react-icons/md";
import {
  showCopyMenu,
  showDeleteMenu,
  showMoveMenu,
  showTagsMenu,
  showAttachmentMenu,
} from "../../../features/modalActionMenusVisibilitySlice";
import CreateTagMenu from "./TagsMenu/CreateTagMenu/CreateTagMenu";
import TagsMenu from "./TagsMenu/TagsMenu";
import CopyMenu from "./CopyMenu/CopyMenu";
import MoveMenu from "./MoveMenu/MoveMenu";
import DeleteMenu from "./DeleteMenu/DeleteMenu";
import AttachmentMenu from "./AttachmentMenu/AttachmentMenu";
import { useAppDispatch, useAppSelector } from "../../../app/hooks";

const CardModalActions = () => {
  const dispatch = useAppDispatch();
  const tagsMenuOpen = useAppSelector((state) => state.modalActionMenusVisibility.value.tagsMenuOpen);
  const createTagsMenuOpen = useAppSelector((state) => state.modalActionMenusVisibility.value.createTagsMenuOpen);
  const copyMenuOpen = useAppSelector((state) => state.modalActionMenusVisibility.value.copyMenuOpen);
  const moveMenuOpen = useAppSelector((state) => state.modalActionMenusVisibility.value.moveMenuOpen);
  const deleteMenuOpen = useAppSelector((state) => state.modalActionMenusVisibility.value.deleteMenuOpen);
  const attachmentMenuOpen = useAppSelector((state) => state.modalActionMenusVisibility.value.attachmentMenuOpen);

  const handleShowTagsMenu = () => {
    dispatch(showTagsMenu());
  };

  const handleShowCopyMenu = () => {
    dispatch(showCopyMenu());
  };

  const handleShowMoveMenu = () => {
    dispatch(showMoveMenu());
  };

  const handleShowDeleteMenu = () => {
    dispatch(showDeleteMenu());
  };

  const handleShowAttachmentMenu = (id: number) => {
    dispatch(showAttachmentMenu(id));
  };

  return (
    <div className="flex flex-col w-44 py-4 ml-auto">
      <h4 className="text-gray-800 text-sm">Add to card</h4>
      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowTagsMenu}
        >
          <AiOutlineTag />
          <p>Tags</p>
        </button>
        {tagsMenuOpen && <TagsMenu />}
      </div>
      {createTagsMenuOpen && <CreateTagMenu />}
      <h4 className="text-gray-800 text-sm">Actions</h4>
      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowCopyMenu}
        >
          <MdOutlineContentCopy />
          <p>Copy</p>
        </button>
        {copyMenuOpen && <CopyMenu />}
      </div>

      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowMoveMenu}
        >
          <HiOutlineArrowRight />
          <p>Move</p>
        </button>
        {moveMenuOpen && <MoveMenu />}
      </div>

      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={handleShowDeleteMenu}
        >
          <AiOutlineDelete />
          <p>Delete</p>
        </button>
        {deleteMenuOpen && <DeleteMenu />}
      </div>

      <div className="flex-col flex">
        <button
          className="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
          onClick={() => handleShowAttachmentMenu(1)}
        >
          <AiOutlinePaperClip />
          <p>Attachment</p>
        </button>
        {attachmentMenuOpen.status && attachmentMenuOpen.id === 1 && <AttachmentMenu update={false} />}
      </div>
    </div>
  );
};

export default CardModalActions;
