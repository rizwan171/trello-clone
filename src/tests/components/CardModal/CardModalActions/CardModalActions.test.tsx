import { fireEvent, screen } from "@testing-library/react";
import CardModalActions from "../../../../components/CardModal/CardModalActions/CardModalActions";
import { renderWithProviders } from "../../../utils/renderUtils";

describe("CardModalActions", () => {
  it("should render successfully", () => {
    const view = renderWithProviders(<CardModalActions />);
    expect(view.asFragment()).toMatchSnapshot();
  });

  it("should show tags menu", () => {
    const { store } = renderWithProviders(<CardModalActions />);
    expect(store.getState().modalActionMenusVisibility.value.tagsMenuOpen).toBeFalsy();

    const tagsMenuButton = screen.getByTestId("card-modal-actions-tags-button") as HTMLButtonElement;
    expect(tagsMenuButton).toBeInTheDocument();

    fireEvent.click(tagsMenuButton);

    expect(store.getState().modalActionMenusVisibility.value.tagsMenuOpen).toBeTruthy();
  });

  it("should show copy menu", () => {
    const { store } = renderWithProviders(<CardModalActions />);
    expect(store.getState().modalActionMenusVisibility.value.copyMenuOpen).toBeFalsy();

    const copyMenuButton = screen.getByTestId("card-modal-actions-copy-button") as HTMLButtonElement;
    expect(copyMenuButton).toBeInTheDocument();

    fireEvent.click(copyMenuButton);

    expect(store.getState().modalActionMenusVisibility.value.copyMenuOpen).toBeTruthy();
  });

  it("should show move menu", () => {
    const { store } = renderWithProviders(<CardModalActions />);
    expect(store.getState().modalActionMenusVisibility.value.moveMenuOpen).toBeFalsy();

    const moveMenuButton = screen.getByTestId("card-modal-actions-move-button") as HTMLButtonElement;
    expect(moveMenuButton).toBeInTheDocument();

    fireEvent.click(moveMenuButton);

    expect(store.getState().modalActionMenusVisibility.value.moveMenuOpen).toBeTruthy();
  });

  it("should show delete menu", () => {
    const { store } = renderWithProviders(<CardModalActions />);
    expect(store.getState().modalActionMenusVisibility.value.deleteMenuOpen).toBeFalsy();

    const deleteMenuButton = screen.getByTestId("card-modal-actions-delete-button") as HTMLButtonElement;
    expect(deleteMenuButton).toBeInTheDocument();

    fireEvent.click(deleteMenuButton);

    expect(store.getState().modalActionMenusVisibility.value.deleteMenuOpen).toBeTruthy();
  });

  it("should show attachment menu", () => {
    const { store } = renderWithProviders(<CardModalActions />);
    expect(store.getState().modalActionMenusVisibility.value.attachmentMenuOpen.id).toBe(0);
    expect(store.getState().modalActionMenusVisibility.value.attachmentMenuOpen.status).toBeFalsy();

    const attachmentMenuButton = screen.getByTestId("card-modal-actions-attachment-menu-button") as HTMLButtonElement;
    expect(attachmentMenuButton).toBeInTheDocument();

    fireEvent.click(attachmentMenuButton);

    expect(store.getState().modalActionMenusVisibility.value.attachmentMenuOpen.id).toBe(1);
    expect(store.getState().modalActionMenusVisibility.value.attachmentMenuOpen.status).toBeTruthy();
  });
});
