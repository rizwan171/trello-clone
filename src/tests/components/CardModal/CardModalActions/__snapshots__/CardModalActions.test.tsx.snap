// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardModalActions should render successfully 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col w-44 py-4 ml-auto"
  >
    <h4
      class="text-gray-800 text-sm"
    >
      Add to card
    </h4>
    <div
      class="flex-col flex"
    >
      <button
        class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
        data-testid="card-modal-actions-tags-button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M938 458.8l-29.6-312.6c-1.5-16.2-14.4-29-30.6-30.6L565.2 86h-.4c-3.2 0-5.7 1-7.6 2.9L88.9 557.2a9.96 9.96 0 0 0 0 14.1l363.8 363.8c1.9 1.9 4.4 2.9 7.1 2.9s5.2-1 7.1-2.9l468.3-468.3c2-2.1 3-5 2.8-8zM459.7 834.7L189.3 564.3 589 164.6 836 188l23.4 247-399.7 399.7zM680 256c-48.5 0-88 39.5-88 88s39.5 88 88 88 88-39.5 88-88-39.5-88-88-88zm0 120c-17.7 0-32-14.3-32-32s14.3-32 32-32 32 14.3 32 32-14.3 32-32 32z"
          />
        </svg>
        <p>
          Tags
        </p>
      </button>
    </div>
    <h4
      class="text-gray-800 text-sm"
    >
      Actions
    </h4>
    <div
      class="flex-col flex"
    >
      <button
        class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
        data-testid="card-modal-actions-copy-button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
          />
        </svg>
        <p>
          Copy
        </p>
      </button>
    </div>
    <div
      class="flex-col flex"
    >
      <button
        class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
        data-testid="card-modal-actions-move-button"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M14 5l7 7m0 0l-7 7m7-7H3"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <p>
          Move
        </p>
      </button>
    </div>
    <div
      class="flex-col flex"
    >
      <button
        class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
        data-testid="card-modal-actions-delete-button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
          />
        </svg>
        <p>
          Delete
        </p>
      </button>
    </div>
    <div
      class="flex-col flex"
    >
      <button
        class="flex gap-2 py-1 px-2 mb-2 bg-trello-gray-card-modal-buttons hover:bg-trello-gray-card-modal-buttons-hover text-trello-blue-card-modal-button-text items-center text-base shadow-sm rounded-sm"
        data-testid="card-modal-actions-attachment-menu-button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M779.3 196.6c-94.2-94.2-247.6-94.2-341.7 0l-261 260.8c-1.7 1.7-2.6 4-2.6 6.4s.9 4.7 2.6 6.4l36.9 36.9a9 9 0 0 0 12.7 0l261-260.8c32.4-32.4 75.5-50.2 121.3-50.2s88.9 17.8 121.2 50.2c32.4 32.4 50.2 75.5 50.2 121.2 0 45.8-17.8 88.8-50.2 121.2l-266 265.9-43.1 43.1c-40.3 40.3-105.8 40.3-146.1 0-19.5-19.5-30.2-45.4-30.2-73s10.7-53.5 30.2-73l263.9-263.8c6.7-6.6 15.5-10.3 24.9-10.3h.1c9.4 0 18.1 3.7 24.7 10.3 6.7 6.7 10.3 15.5 10.3 24.9 0 9.3-3.7 18.1-10.3 24.7L372.4 653c-1.7 1.7-2.6 4-2.6 6.4s.9 4.7 2.6 6.4l36.9 36.9a9 9 0 0 0 12.7 0l215.6-215.6c19.9-19.9 30.8-46.3 30.8-74.4s-11-54.6-30.8-74.4c-41.1-41.1-107.9-41-149 0L463 364 224.8 602.1A172.22 172.22 0 0 0 174 724.8c0 46.3 18.1 89.8 50.8 122.5 33.9 33.8 78.3 50.7 122.7 50.7 44.4 0 88.8-16.9 122.6-50.7l309.2-309C824.8 492.7 850 432 850 367.5c.1-64.6-25.1-125.3-70.7-170.9z"
          />
        </svg>
        <p>
          Attachments
        </p>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
